[plugin] Add PHP grammar contributions #4582

vrubezhny opened the pull request:
The CQs [1], [2] are resolved, so it's possible to contribute PHP grammars
Need to remove the similar contribution from PHP extension[3]

[1] https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19282
[2] https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19331
[3] https://github.com/theia-ide/theia-php-extension

Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

akosyakov commented on packages/textmate-grammars/src/browser/phphtml.ts:
```diff
@@ -0,0 +1,93 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
```
it's 2019

vrubezhny replied:
yes, sorry. I'll add a fixup

akosyakov commented on packages/textmate-grammars/src/browser/phphtml.ts:
```diff
@@ -0,0 +1,93 @@
+    readonly id = 'php';
+    readonly scopeName = 'text.html.php';
+
+    readonly config: monaco.languages.LanguageConfiguration = {
```
if it's copied from MS repo, please attribute copyrights properly

vrubezhny replied:
Technically it's not. 
File packages/textmate-grammars/src/browser/php.ts as well as thisone are the modified copies of https://github.com/theia-ide/theia-php-extension/blob/master/php/src/browser/php-grammar-contribution.ts, not directly copied MS. So, the copyright here probably should be of 'TypeFox' (if not the 'RedHat').

The only doubt I have here is that probably we souldn't have it in this separate file - probably it's better to combine registering the language and HTML PHP grammar inside packages/textmate-grammars/src/browser/php.ts if possible (similar to how it's done in MS's https://github.com/Microsoft/vscode/blob/master/extensions/php/package.json#L32).

akosyakov replied:
> The only doubt I have here...

i see, php language config should be registered only once in `PhpGrammarContribution`, no need to register it again in `PhpHtmlGrammarContribution`.

akosyakov commented on packages/textmate-grammars/src/browser/phphtml.ts:
```diff
@@ -0,0 +1,93 @@
+            }
+        });
+
+        registry.mapLanguageIdToTextmateGrammar(this.id, this.scopeName);
```
I don't thin we support multiple scopes per a language. It looks like we need to drop phphtml support till it is implemented. Could you file an issue to support multiple scopes?

akosyakov replied:
verified, it breaks:
<img width="1737" alt="Screen Shot 2019-04-02 at 06 24 17" src="https://user-images.githubusercontent.com/3082655/55376275-04360e80-5510-11e9-9579-0a155bc2a917.png">

php grammar gets overridden by phphtml, we need to pick one or another scope name for now

vrubezhny replied:
OK. I'll remove phphtml for now and open an issue to support multiple scopes

tolusha commented:
@akosyakov 
CQ with TYPE_A is approved. 
Do we need a CQ with TYPE_B ?

akosyakov commented:
@tolusha Type B should be fine as far as I know cc @marcdumais-work

vrubezhny commented:
The issue https://github.com/theia-ide/theia/issues/4796 is opened for multiple contexts per a language support

akosyakov commented:
We need a PR against https://github.com/theia-ide/theia-php-extension which removes textmate grammars. It should be merged immediately after this PR is merged to avoid breaking clients using next.

vrubezhny commented:
@akosyakov Then I'm going to prepare a PR for https://github.com/theia-ide/theia-php-extension before we merge this one

vrubezhny commented:
The following PR is added in order to remove PHP grammar contribution from`theia-php-extension`: 

https://github.com/theia-ide/theia-php-extension/pull/17

akosyakov commented on packages/textmate-grammars/data/php.html.tmLanguage.json:
```diff
@@ -0,0 +1,189 @@
+{
```
It does not seem to be used, please remove if it not

vrubezhny replied:
But then, when #4796 is resolved, we'll have to add it again. Are your sure we must remove it for now?

vrubezhny replied:
> There should a PR against https://github.com/theia-ide/theia-php-extension which removes textmate grammars. It should be merged immediately after this PR is merged to avoid breaking clients using next.
> 
> Code looks good now in this PR.

I can merge this PR, but it looks like I'm not allowed to merge complimentary grammars removal PR https://github.com/theia-ide/theia-php-extension/pull/17 

Could you, please, merge that complimentary PR once we've merged this one?

akosyakov replied:
sure i can do, just ping me there after merging

akosyakov replied:
> But then, when #4796 is resolved, we'll have to add it again. Are your sure we must remove it for now?

you can comment on the issue that it should be added after the PR is resolved, right now i don't see a reason to commit it

vrubezhny replied:
OK. Then I'm removing it from the current PR.

vrubezhny replied:
I've removed packages/textmate-grammars/data/php.html.tmLanguage.json from the PR. So it looks like we only need to wait for the CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19509 to be resolved

vrubezhny commented:
@akosyakov I've squashed the fixup in PR and rebased it against current master. (however currently unused packages/textmate-grammars/data/php.html.tmLanguage.json  is still not excluded from the PR)

marcdumais-work commented:
Hi @tolusha 

> CQ with TYPE_A is approved.
> Do we need a CQ with TYPE_B ?

For cope we copy from a non-eclipse project and (potentially) modify, I think we need a "Type B" CQ. 

So I do not think the [CQ you opened](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19331) is valid for this case. On the first CQ submission page, you should chose:
![image](https://user-images.githubusercontent.com/25749063/55626727-4bbfd300-577a-11e9-9bf1-39483a457fba.png)

More detailed instructions for this case are [here](https://github.com/theia-ide/theia/wiki/Registering-CQs#case-3rd-party-project-code-copiedforked-from-another-project-into-eclipse-theia-maintained-by-us)

vrubezhny commented:
@tolusha, @marcdumais-work , @akosyakov  Yet another CQ is opened for the issue, so we need to wait again for it to be resolved.

akosyakov commented:
@vrubezhny please link a CQ

vrubezhny commented:
@akosyakov Sorry, I forgot the link:
The CQ is created for this issue: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19509

akosyakov commented:
@vrubezhny CQ is good: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19509#c3 please rebase and merge
